assets/js/modules/analytics-4/utils/partition-report.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * Analytics-4 partition report utility.
+ *
+ * Site Kit by Google, Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Splits report data into segments for current and comparison ranges.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Array}  report               Report rows.
+ * @param {Object} args                 Additional arguments.
+ * @param {number} args.dateRangeLength Date range length of report segments.
+ * @return {Object} Object with keys for `compareRange` and `currentRange`.
+ */
+export function partitionAnalytics4Report( report, { dateRangeLength } ) {
```

Comment:
Unit tests for this function and the `extractAnalytics4DashboardData` function can be implemented in a follow-up ticket due to the overestimate.